core/src/relabeling/LLRelabelingStrategy.cpp

```diff
@@ -0,0 +1,82 @@
+/*-------------------------------------------------------------------------------
+  This file is part of generalized random forest (grf).
+
+  grf is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  grf is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with grf. If not, see <http://www.gnu.org/licenses/>.
+ #-------------------------------------------------------------------------------*/
+
+#include "relabeling/LLRelabelingStrategy.h"
+
+namespace grf {
+
+LLRelabelingStrategy::LLRelabelingStrategy(double split_lambda,
+                                           bool weight_penalty,
+                                           std::vector<size_t> ll_split_variables):
+  split_lambda(split_lambda),
+  weight_penalty(weight_penalty),
+  ll_split_variables(ll_split_variables){
+};
+
+bool LLRelabelingStrategy::relabel(
+    const std::vector<size_t>& samples,
+    const Data& data,
+    std::vector<double>& responses_by_sample) const {
+
+  size_t num_variables = ll_split_variables.size();
+  size_t num_data_points = samples.size();
+
+  Eigen::MatrixXd X (num_data_points, num_variables+1);
+  Eigen::MatrixXd Y (num_data_points, 1);
+  for (size_t i = 0; i < num_data_points; ++i) {
+    for (size_t j = 0; j < num_variables; ++j){
+      size_t current_predictor = ll_split_variables[j];
+      X(i,j+1) = data.get(samples[i],current_predictor);
+    }
+    Y(i) = data.get_outcome(samples[i]);
+    X(i, 0) = 1;
+  }
+
+  // find ridge regression predictions
+  Eigen::MatrixXd M_unpenalized(num_variables+1, num_variables+1);
+  M_unpenalized.noalias() = X.transpose()*X;
+
+  Eigen::MatrixXd M = M_unpenalized;
+  // M = M_unpenalized;
+
+  if (!weight_penalty) {
+    // standard ridge penalty
+    double normalization = M.trace() / (num_variables + 1);
+    for (size_t j = 1; j < num_variables + 1; ++j){
+      M(j,j) += split_lambda * normalization;
+    }
+  } else {
+    // covariance ridge penalty
+    for (size_t j = 1; j < num_variables+1; ++j){
+      M(j,j) += split_lambda * M(j,j); // note that the weights are already normalized
+    }
+  }
+
+  Eigen::MatrixXd local_coefficients = M.ldlt().solve(X.transpose()*Y);
+  Eigen::MatrixXd leaf_predictions = X.transpose()*local_coefficients;
```

Comment:
X should not be transposed. If you compile the package with `-UNDEBUG` the Eigen assertions will warn you (`Assertion failed: (lhs.cols() ==rhs.rows() && "invalid matrix product"...`)

Reply:
I'll push a commit to fix this today! Thanks :)

Reply:
Can you also go over your tests again and make sure they seem fine, it is not very confidence instilling that they all passed with this error

Reply:
Yup there are another few tests already that I want to add to this PR, I'll ping you when I think they are more comprehensive